Stellar-transaction.x

```diff
@@ -474,23 +474,87 @@ struct LiquidityPoolWithdrawOp
     int64 minAmountB; // minimum amount of second asset to withdraw
 };
 
-enum HostFunction
+enum HostFunctionType
 {
-    HOST_FN_INVOKE_CONTRACT = 0,
-    HOST_FN_CREATE_CONTRACT_WITH_ED25519 = 1,
-    HOST_FN_CREATE_CONTRACT_WITH_SOURCE_ACCOUNT = 2,
-    HOST_FN_CREATE_TOKEN_CONTRACT_WITH_SOURCE_ACCOUNT = 3,
-    HOST_FN_CREATE_TOKEN_CONTRACT_WITH_ASSET = 4
+    HOST_FUNCTION_TYPE_INVOKE_CONTRACT = 0,
+    HOST_FUNCTION_TYPE_CREATE_CONTRACT = 1,
+    HOST_FUNCTION_TYPE_INSTALL_CONTRACT_CODE = 2
+};
+
+enum ContractIDType
+{
+    CONTRACT_ID_FROM_PUBLIC_KEY = 0,
+    CONTRACT_ID_FROM_ASSET = 1
+};
+ 
+enum ContractIDPublicKeyType
+{
+    CONTRACT_ID_PUBLIC_KEY_SOURCE_ACCOUNT = 0,
+    CONTRACT_ID_PUBLIC_KEY_ED25519 = 1
+};
```

Comment:
This relationship feels a bit backwards. It suggests a source account is a public key, when a source account contains a public key. Could "source account" be a `ContractIDType` variant instead? Then you don't need as much nesting. Nesting comes at a XDR size cost, and hierarchies make more assumptions about the future. It doesn't appear to offer a ton of value in this case. Am I missing something of value there?

```suggestion
enum ContractIDType
{
    CONTRACT_ID_FROM_SOURCE_ACCOUNT = 0,
    CONTRACT_ID_FROM_PUBLIC_KEY_ED25519 = 1,
    CONTRACT_ID_FROM_ASSET = 2
};
```

Reply:
I use nesting for the sake of avoiding duplication and to represent the logical structure of the data. Any kind of id inherited from a public key would contain salt, but the source of the public key itself might be different. If we add one more type of a public key in the future, it would just go into `ContractIDPublicKeyType` enum and will by definition have the salt.

I'm not feeling too strongly about that though and I agree the difference is pretty marginal. If that apparently doesn't improve clarity, we can remove the nesting as well.

Reply:
>Any kind of id inherited from a public key would contain salt

This is an assumption. I don't think we know if it would be true for all ways we create contracts from public keys.